doc/gamecomp.1
.\" t
.de TQ
.  br
.  ns
.  TP \\$1
..
.TH GAMEARCH "1" "November 2010" "Camoto" "User Commands"
.SH NAME
gamecomp \- apply or reverse compression and encryption algorithms used by
games
.SH SYNOPSIS
.B gamecomp
\fB--type=\fR\fItype\fR [\fB--apply\fR] < \fIinfile\fR > \fIoutfile\fR
.PP
.B gamecomp
\fB--list\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Filter data from standard input and send it to standard output, using the
algorithm given by the \fB--type\fR parameter.  The default is to reverse the
algorithm (which will decompress or decrypt the input data) however the
\fB--apply\fR option will apply the algorithm instead (resulting in the input
data being compressed or encrypted.)
.SH OPTIONS
.TP
\fB-l\fR, \fB--list\fR
list all the algorithms on stdout instead
.TP
\fB-t\fR, \fB--type\fR
specify the algorithm to use, from the list shown by \fB--list\fR
.SH "EXAMPLES"
.TP
\fBgamearch bash1.dat -u -x mb1_h1.txt\fR
.TQ
\fBgamecomp -t lzw-bash < mb1_h1.txt > decompressed.txt\fR
extract mb1_h1.txt from the Monster Bash archive file (-x), without
decompressing it (-u).  Then use \fBgamecomp\fR to decompress it
instead, saving the data into decompressed.txt.
.TP
\fBgamecomp -a -t lzw-bash < input.txt > mb1_h1.txt\fR
take the plain text in input.txt, compress it using the Monster Bash
compression algorithm and store the compressed data in mb1_h1.txt.
.SH NOTES
.PP
Exit status is 0 on success, 1 on bad parameters and 2 on an I/O error.
.SH KNOWN ISSUES
.PP
Not all algorithms support compression.
.SH AUTHOR
Written by Adam Nielsen <malvineous@shikadi.net>.
.SH "REPORTING BUGS"
Report bugs to <malvineous@shikadi.net>.
.SH COPYRIGHT
Copyright \(co 2010 Adam Nielsen.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
\fBgamearch\fR(1), <http://www.shikadi.net/camoto>
.PP
All the algorithms supported by this utility are fully documented on the
Game Modding Wiki \fB<http://www.shikadi.net/moddingwiki>\fR.
.PP
This utility is part of the Camoto suite of tools.


examples/gamecomp.cpp
/**
 * @file   gamecomp.cpp
 * @brief  Command-line interface to the compression/encryption filters in
 *         libgamearchive.
 *
 * Copyright (C) 2010 Adam Nielsen <malvineous@shikadi.net>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <boost/algorithm/string.hpp> // for case-insensitive string compare
#include <boost/program_options.hpp>
#include <boost/filesystem.hpp>
#include <boost/iostreams/copy.hpp>
#include <boost/bind.hpp>
#include <camoto/gamearchive.hpp>
#include <camoto/util.hpp>
#include <iostream>
#include <fstream>

namespace po = boost::program_options;
namespace fs = boost::filesystem;
namespace ga = camoto::gamearchive;

#define PROGNAME "gamecomp"

// Return values
#define RET_OK           0  ///< All is good
#define RET_BADARGS      1  ///< Bad arguments (missing/invalid parameters)
#define RET_SHOWSTOPPER  2  ///< I/O error

/// Delete function to allow static object to be passed to a shared_ptr
struct null_deleter
{
	void operator()(void const *) const { }
};

int main(int iArgC, char *cArgV[])
{
	// Set a better exception handler
	std::set_terminate( __gnu_cxx::__verbose_terminate_handler );

	// Disable stdin/printf/etc. sync for a speed boost
	std::ios_base::sync_with_stdio(false);

	// Declare the supported options.
	po::options_description poOptions("Options");
	poOptions.add_options()
		("apply,a",
			"apply the filter instead (compress/encrypt the input data) rather than "
			"the default of reversing the filter (to decompress/decrypt)")
		("list,l",
			"list all filters")
		("type,t", po::value<std::string>(),
			"specify the filter type")
	;

	po::options_description poHidden("Hidden parameters");
	poHidden.add_options()
		("help", "produce help message")
	;

	po::options_description poVisible("");
	poVisible.add(poOptions);

	po::options_description poComplete("Parameters");
	poComplete.add(poOptions).add(poHidden);
	po::variables_map mpArgs;

	ga::ManagerPtr pManager(ga::getManager());

	ga::FilterTypePtr pFilterType;
	bool bApply = false;   // default is to reverse the algorithm (decompress)
	try {
		po::parsed_options pa = po::parse_command_line(iArgC, cArgV, poComplete);

		// Parse the global command line options
		for (std::vector<po::option>::iterator i = pa.options.begin(); i != pa.options.end(); i++) {
			if (i->string_key.empty()) {
				std::cerr << "Error: unexpected extra parameter (you can't list "
					"filenames on the command line, you must redirect stdin/out)"
					<< std::endl;
				return RET_BADARGS;
			} else if (i->string_key.compare("help") == 0) {
				std::cout <<
					"Copyright (C) 2010 Adam Nielsen <malvineous@shikadi.net>\n"
					"This program comes with ABSOLUTELY NO WARRANTY.  This is free software,\n"
					"and you are welcome to change and redistribute it under certain conditions;\n"
					"see <http://www.gnu.org/licenses/> for details.\n"
					"\n"
					"Utility to apply and reverse compression and encryption algorithms used by\n"
					"games on their data files.\n"
					"Build date " __DATE__ " " __TIME__ << "\n"
					"\n"
					"Usage: gamecomp -t <type> < infile > outfile\n" << poVisible << "\n"
					<< std::endl;
				return RET_OK;
			} else if (
				(i->string_key.compare("t") == 0) ||
				(i->string_key.compare("type") == 0)
			) {
				if (i->value.size() == 0) {
					std::cerr << PROGNAME ": --type (-t) requires a parameter."
						<< std::endl;
					return RET_BADARGS;
				}
				pFilterType = pManager->getFilterTypeByCode(i->value[0]);
				if (!pFilterType) {
					std::cerr << PROGNAME ": Unknown filter code given by --type (-t) - "
						"use -l for a list." << std::endl;
					return RET_BADARGS;
				}
			} else if (
				(i->string_key.compare("a") == 0) ||
				(i->string_key.compare("apply") == 0)
			) {
				bApply = true;
			} else if (
				(i->string_key.compare("l") == 0) ||
				(i->string_key.compare("list") == 0)
			) {
				for (int i = 0; ; i++) {
					ga::FilterTypePtr pFilterType(pManager->getFilterType(i));
					if (!pFilterType) break;
					std::cout << pFilterType->getFilterCode() << std::endl;
				}
				return RET_OK;
			}
		}

		if (!pFilterType) {
			std::cerr << PROGNAME ": No filter type given (--type/-t).  Use -l for "
				"a list." << std::endl;
			return RET_BADARGS;
		}

		// Convert std::cin into a shared pointer
		camoto::istream_sptr stdin(&std::cin, null_deleter());

		// Apply the filter
		camoto::istream_sptr in(pFilterType->apply(stdin));

		// Copy filtered data to stdout
		boost::iostreams::copy(*in, std::cout);

	} catch (const std::ios::failure& e) {
		std::cerr << PROGNAME ": I/O error - " << e.what() << std::endl;
		return RET_SHOWSTOPPER;
	} catch (po::unknown_option& e) {
		std::cerr << PROGNAME ": " << e.what()
			<< ".  Use --help for help." << std::endl;
		return RET_BADARGS;
	} catch (po::invalid_command_line_syntax& e) {
		std::cerr << PROGNAME ": " << e.what()
			<< ".  Use --help for help." << std::endl;
		return RET_BADARGS;
	} catch (const camoto::ECorruptedData& e) {
		std::cout << std::flush; // keep as much data as we could process
		std::cerr << PROGNAME ": Decompression failed.  " << e.what() << std::endl;
		return RET_SHOWSTOPPER;
	}

	return RET_OK;
}
